src/ffi/types.rs

```diff
@@ -0,0 +1,1612 @@
+	ConfirmationStatus, LSPFeeLimits, PaymentDirection, PaymentKind, PaymentStatus,
+};
+pub use crate::payment::{MaxTotalRoutingFeeLimit, QrPaymentResult, SendingParameters};
+
+pub use lightning::chain::channelmonitor::BalanceSource;
+pub use lightning::events::{ClosureReason, PaymentFailureReason};
+pub use lightning::ln::types::ChannelId;
+pub use lightning::offers::offer::OfferId;
+pub use lightning::routing::gossip::{NodeAlias, NodeId, RoutingFees};
+pub use lightning::util::string::UntrustedString;
+
+pub use lightning_types::payment::{PaymentHash, PaymentPreimage, PaymentSecret};
+
+pub use lightning_invoice::{Description, SignedRawBolt11Invoice};
+
+pub use lightning_liquidity::lsps1::msgs::ChannelInfo as ChannelOrderInfo;
+pub use lightning_liquidity::lsps1::msgs::{OrderId, OrderParameters, PaymentState};
+
+pub use bitcoin::{Address, BlockHash, FeeRate, Network, OutPoint, Txid};
+
+pub use bip39::Mnemonic;
+
+pub use vss_client::headers::{VssHeaderProvider, VssHeaderProviderError};
+
+pub type DateTime = chrono::DateTime<chrono::Utc>;
+
+use crate::UniffiCustomTypeConverter;
+
+use crate::builder::sanitize_alias;
+use crate::error::Error;
+use crate::hex_utils;
+use crate::{SocketAddress, UserChannelId};
+
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::hashes::Hash;
+use bitcoin::secp256k1::PublicKey;
+use lightning::ln::channelmanager::PaymentId;
+use lightning::offers::invoice::Bolt12Invoice as LdkBolt12Invoice;
+use lightning::offers::offer::{Amount as LdkAmount, Offer as LdkOffer};
+use lightning::offers::refund::Refund as LdkRefund;
+use lightning::util::ser::Writeable;
+use lightning_invoice::{Bolt11Invoice as LdkBolt11Invoice, Bolt11InvoiceDescriptionRef};
+
+use std::convert::TryInto;
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+
+impl UniffiCustomTypeConverter for PublicKey {
+	type Builtin = String;
+
+	fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+		if let Ok(key) = PublicKey::from_str(&val) {
+			return Ok(key);
+		}
+
+		Err(Error::InvalidPublicKey.into())
+	}
+
+	fn from_custom(obj: Self) -> Self::Builtin {
+		obj.to_string()
+	}
+}
+
+impl UniffiCustomTypeConverter for NodeId {
+	type Builtin = String;
+
+	fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+		if let Ok(key) = NodeId::from_str(&val) {
+			return Ok(key);
+		}
+
+		Err(Error::InvalidNodeId.into())
+	}
+
+	fn from_custom(obj: Self) -> Self::Builtin {
+		obj.to_string()
+	}
+}
+
+impl UniffiCustomTypeConverter for Address {
+	type Builtin = String;
+
+	fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+		if let Ok(addr) = Address::from_str(&val) {
+			return Ok(addr.assume_checked());
+		}
+
+		Err(Error::InvalidAddress.into())
+	}
+
+	fn from_custom(obj: Self) -> Self::Builtin {
+		obj.to_string()
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum OfferAmount {
+	Bitcoin { amount_msats: u64 },
+	Currency { iso4217_code: String, amount: u64 },
+}
```

Comment:
Curious about why `LdkAmount` is not used directly here. `OfferAmount` is basically the same struct except for the `iso4217_code` field. Is this because there is no support for fixed-sized arrays?